### ISSUE TITLE
Support for polygon outlines on terrain

### ISSUE THREAD

**Issue by ediebold** [CesiumGS/cesium#6694]
I saw the exciting development in https://github.com/AnalyticalGraphicsInc/cesium/pull/6615 that will allow polylines on terrain!

I was wondering if this new method could also be used for polygon outlines? I assume those are drawn in a very similar way. If we can get polygon outlines on terrain, I believe Cesium will fully support clamping all the classic vector data types to terrain!

**Comment by hpinkos** [comment]
Hi @ediebold!  While that polylines on terrain PR gets us closer to supporting this, there's a little bit more work we'd have to do to make sure the outlines match up with the fill for all of the ground geometry types (corridor, ellipse, polygon and rectangle).  I'm just not sure how soon we'll have a chance to look into it, but I don't think it will take very much time to do.  Thanks for opening this issue!  We'll give you an update when we have one =)

**Comment by mramato** [comment]
We've talked in the past about replacing the standard WebGL outlines we currently use with polyline primitive lines.  The benefit to this is that they would be stylable and as of this month, would support clamping on terrain.  @bagnell said the change itself is not very hard but there was concern about outline quality (awkward corners of a box for example).

Of course we haven't had this conversation in years and I definitely think it's time to revisit it.  I don't know how we would get to "Terrain on by default" without it.

**Comment by bagnell** [comment]
Polylines on terrain would have an issue where the lines would overlap the polygon because of the constant pixel with. This might only be a problem for translucent polylines though. The other issue is with extruded geometry where the lines from the top surface to the extruded bottom wouldn't connect. Here is an example:
![image](https://user-images.githubusercontent.com/1494815/41688919-1303321a-74bc-11e8-9207-1aa27ad27bc2.png)

**Comment by mramato** [comment]
I'm curious as to what that looks like with a much narrower outlines, (1-2 pixels).

With the number of posts we get about "outline width not working" and the necessity of outlines on terrain in order to support terrain on by default, my vote would be to make this change.  I think it's a net positive win and people would be more than willing to accept the artifacts if they can have terrain and line width.  We can always look at improving these lines down the line when we have more time.

Anyone else want to provide any thoughts here?

**Comment by bagnell** [comment]
Same issues but not as noticeable at 2 pixels.
![image](https://user-images.githubusercontent.com/1494815/41689600-4d5847e0-74bf-11e8-92ab-201214313704.png)

**Comment by ediebold** [comment]
Speaking as a random user, I think the number of use cases this would add support to will outweigh the ones it would won't. I think most people would still be happy with this update even if it came with some caveats such as transparent or thick lines looking weird.

**Comment by mramato** [comment]
@bagnell I think it's important to note that there are already existing artifacts with outlines on polygons, so this might actually improve things in that department.  Here's the GeoJSON example with outlines turned on, notice how the lines are broken up even though the extruded polygon is translucent:

![image](https://user-images.githubusercontent.com/1495196/41719140-2a4b6df4-752d-11e8-9e30-af7fc1fdca3f.png)

[Modified Sandcastle demo](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=dVXbbtNAEP2VIS9x1WqdPvAASStQWloQtKhB8NCgamNP4oX1rrWXhFD135m117GTFitJ5Z0zZ25np2tuYC1wgwbOQOEGpmiFL9n3+iwZZvXrVCvHhUIzPBrP1Vyl6QwxB1cgGK5yXYLy5YI4Vkgg7rSBJX0NVsgdX0iCofXSWTZXMcAX7gpm0d3V/je1eyBNRk2INSVWGV0Ki5RZdLpC/clqdUGkM+1NhkxqnidDxtLd531V2XTGy0piwKUKH05H5YO3D9Zxh5Y5XelfxFLXEkMwKkUlS68yJ7RK8l2AI3icK6CnaRLrLJbxPO8j67QDNE2v0NXN4cbwLegloHLCCbSRi2prT6i4joO1p2zNpUe7owwemZbaXHNbkMvj07gxhDYnwSrodDSmP5MdNZOoVq6gw+PjXR1Ngh/IDXlWNNjtCWSGJkUJt/Osg8GCWxqzVnUxdftA8RJZn2qmy2izUPA1QkmDFtT9XR4nYDVsAkgbrKkadqGA95mKUBtBXcFdwHvboC2FjC6h3HAUqE0M28tm19gtdaMNfy9+jvchoYYWsGXh7QDQBDvrWn4fQH0asYTkVW3e62x4Wueo2Wl4ZUvSWSP15AAeHi6rgsNbGLE3+8ano/EL5F1GbY491FOrmTieKMVKy+2KJllSx4zgEpwG0tzevHudjL2JXmzn1cXrB7nDUq+bYWnvJG0Ku6eRA7aIIbIll7bf/JeB05c6+uP647fLgzwu/zjjc9wr+LmGh5aMlZc83HUGcNldhD5ZLVfbcBldoamva6vAsIpmtzewpHvjSYqCJMttvEc5hHnvjUGorM2rDQ3C0oUr/Arj+jwJqs/FWlAJiy28Hv1/INhUml+jWBWuE3OXKfvaBUoD1yj2mQTydMQ05WI2tPq6rYfGkJwf2yV2IWwlaX1xtQUykc1SlEx7RVqgEjeFoEseFrBQq5jpRpCaNoxLyiHyjUM4+hmcDCbWbSWeN9B3oqy0ceCNTGhtO6SNHZZIuvDZb3Qssza4TdLWaUKdAZGfzQcH/5TmA8gkt5YsSy/lTPzF+eB8khJ+zy2mertGQ3UFSHF6/jnmz9gkpdfnXk5rueCmx/gP)

**Comment by mramato** [comment]
@lilleyse @likangning93 @ggetz anyone have any thoughts here, either technical or on the desires of the community?

**Comment by likangning93** [comment]
> Polylines on terrain would have an issue where the lines would overlap the polygon because of the constant pixel with

Should be pretty easy to add a flag that lets polylines on terrain only "extend" in one direction for the simulated screen space width thing. Might not even have to be a batch breaker.

For funny corners with regular polylines, I wonder how much more acceptable it'll look if we create polyline loops for the extrusion caps. Maybe want to construct the loops such that the start/end aren't at a corner.

**Comment by mramato** [comment]
@bagnell @likangning93 I think we'll need to think about making this switch for our Oct 1 release as part of our efforts to have terrain on by default.  Since you both seem to have ideas for making this switch, maybe an early PR just to get the discussion going for real?

**Comment by hpinkos** [comment]
Also reported in https://github.com/CesiumGS/cesium/issues/8364

**Comment by pyshx** [comment]
Any changes on this?

**Comment by ethanchristensen01** [comment]
I'd really like to see this get done. Is there anything specific I can do to contribute?

**Comment by ggetz** [comment]
Also requested in https://github.com/CesiumGS/cesium/issues/11438

**Comment by ggetz** [comment]
@ethanchristensen01 If you have the bandwidth to contribution a potential fix, we would be happy to discuss the approach.

Otherwise, use case descriptions and test examples or test data also help as well!

Thanks!

**Comment by ethanchristensen01** [comment]
@ggetz Yes, I can work on this, and I would like to hear the approach you recommend. For now, I'll get started with building Cesium on my machine.

**Comment by ethanchristensen01** [comment]
I found a way to work around this issue temporarily in my own project, but it's not ideal. I would still prefer to contribute a fix here.

Let me know, when you have time, what approach I can start with!

Specifically, it would be good to know:
- How involved is this change?
- Will this change potentially introduce bugs?
- Will I be modifying/creating shader code?

**Comment by javier-lopez-1s** [comment]
I believe this is a very much needed feature **even when not using terrain**, because `clampToGround` seems to be the only way to:
- Prevent rendering issues such as entities being partially clipped (reported [here](https://community.cesium.com/t/czml-polygon-with-outline-not-rendered-properly/37410) and [here](https://community.cesium.com/t/polyline-rendered-inconsistently-partially-below-polygon/37198)).
- Guarantee the order of rendering, that is, data sources with a higher index in the data source collection to be rendered above other data sources. Also guarantee that `zIndex` works for entities within the same data source.
